### DISCUSSION TITLE
How to use [RequiresPreviewFeaturesAttribute] with Razor .cshtml file

### DISCUSSION THREAD

**Discussion by imalkevich** [dotnet/runtime#61399] [Q&A]
Hi Team,

I'm having some preview features that are being used besides other places inside Razor pages.

While adding new attribute to `.cs` files works as it should, I'm having issues with adding `[RequiresPreviewFeatures]` attribute to Razor .cshtml page - actually I can add this attribute, but during the validation I'm having the following error:
```
**Statistics.cshtml**(39,27): Error CA2252: Using 'CpuStatistics' requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information. 

**Pages_Statistics_cshtml.g.cs**(218,16): Error CA2252: 'get_Model' returns the preview type 'StatisticsModel' and therefore needs to opt into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information. 
```

And here is how my recent attempt to add the attribute to Razor page looks like:
![image](https://user-images.githubusercontent.com/7044304/141054754-19bd45f3-4e84-416a-ba31-81239c9ae322.png)

Would please somebody suggest a way how to add the needed attribute to the Razor page?

Thanks

**Comment by danmoseley** [comment]
@buyaa-n maybe you can help here?

**Reply by buyaa-n** [reply]
I have never tried adding an attribute to the razor page, but found [this doc](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-6.0#attribute), seems `@attribute [RequiresPreviewFeatures]` should work (or add full name of the attribute)

**Reply by imalkevich** [reply]
Thanks, but now I have error about generated file:
![image](https://user-images.githubusercontent.com/7044304/141075809-26e8d82b-0e8a-4f41-866f-b1b9abe7dfe8.png)

Is there some option so that auto-generated files are excluded?  Or maybe automatically enriched with this attribute if the source file has it?

Thanks

**Reply by buyaa-n** [reply]
> Is there some option so that auto-generated files are excluded? Or maybe automatically enriched with this attribute if the source file has it?

Sorry I don't know, probably better to ask in the [aspnetcore](https://github.com/dotnet/aspnetcore) repo

**Reply by pranavkm** [reply]
Using @attribute is the correct way to annotate the generated view. Razor generates an assembly level attribute for each view that is compiled in to the assembly. It looks like the preview-feature analyzer does not like that. Here's a minimal repro for the same issue:

```C#
[assembly: System.Reflection.Metadata.MetadataUpdateHandler(typeof(MyApp.Test))]

namespace MyApp;
[System.Runtime.Versioning.RequiresPreviewFeatures]
public class Test {}
```

This results in a build warning. 

That said, I'm not sure if using the `RequiresPreviewFeaturesAttribute` makes a lot of sense. View types aren't referenced by user code (they are referenced by paths and discovered by reflection), so I don't know if adding the attribute necessarily helps.

**Reply by imalkevich** [reply]
Thanks @pranavkm. This means it is a blocking issue for using this feature, right?

**Reply by imalkevich** [reply]
Team,

One thing I want to confirm as I have tried it and have been asked multiple times already - is adding `<EnablePreviewFeatures>true</EnablePreviewFeatures>` in .csproj file eliminates the need to add the attribute in all places? For me it doesn't seem to work... 

Thanks

**Reply by imalkevich** [reply]
Hi Team,

Just wanted to share my feedback on this feature:
1. In my .DLL's .csproj file I added the following line `<EnablePreviewFeatures>true</EnablePreviewFeatures>` - but the compiler still asked me to add `[RequiresPreviewFeatures]` attribute
2. In my consuming project .csproj I added `<EnablePreviewFeatures>true</EnablePreviewFeatures>` - but now I'm blocked by the fact that I can't add this `[RequiresPreviewFeatures]` attribute to Razor's autogenerated file

I followed two documents https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md and https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2252, but my understanding that putting `<EnablePreviewFeatures>true</EnablePreviewFeatures>` inside consuming .csproj doesn't work...

Would be nice to get some better understanding how this should work.

Thanks

**Reply by imalkevich** [reply]
Team, 

Quick update:
1. I removed this experimental feature in favor of a simpler yet not as elegant API with `INumber` I was supposed to use
2. Last thing I tried was - in library .csproj file set `<EnablePreviewFeatures>True</EnablePreviewFeatures>` and `<GenerateRequiresPreviewFeaturesAttribute>False</GenerateRequiresPreviewFeaturesAttribute>`. For consuming project I configured `<EnablePreviewFeatures>True</EnablePreviewFeatures>` and it didn't work.

Sad I had to revert this feature, and even sadder that it worked this way for me...

Thanks

**Reply by danmoseley** [reply]
@tannergooding

**Reply by tannergooding** [reply]
There is probably a disconnect between what https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md currently says and what it actually does. The doc says:
> Note: The analyzer doesn't track the `<EnablePreviewFeatures>` property itself, it tracks the attribute [RequiresPreviewFeatures]. By default, setting `<EnablePreviewFeatures>` to `True` will mark the entire assembly with `[RequiresPreviewFeatures]`, but developers can turn this off which means that they now need to manually annotate parts of their code with [RequiresPreviewFeatures].

But, I believe you need to additionally annotate `GenerateRequiresPreviewFeaturesAttribute=true`.

From what I can tell, that matches what the SDK currently has logic for:
* https://github.com/dotnet/sdk/blob/2a515cdbd8f6be1b019ae2c8d7f21952592f0697/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L40
* https://github.com/dotnet/sdk/blob/2a515cdbd8f6be1b019ae2c8d7f21952592f0697/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L118-L119
* https://github.com/dotnet/sdk/blob/aac033b9324f687eda51d9d7291537bb46cf3a55/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Common.targets#L51-L53

Perhaps @terrajobst can confirm what the intent here was.

**Reply by terrajobst** [reply]
Let's step back. Why would you want to add that attribute to a Razor page in the first place?

The 90% usage scenario is: you enable preview features for your projects, which makes them available everywhere in your project, including Razor pages.

> 1. In my .DLL's .csproj file I added the following line `<EnablePreviewFeatures>true</EnablePreviewFeatures>` - but the compiler still asked me to add `[RequiresPreviewFeatures]` attribute

That would be a bug IMHO, unless you explicitly set `GenerateRequiresPreviewFeaturesAttribute` to `false`.

> 2\. Last thing I tried was - in library .csproj file set `<EnablePreviewFeatures>True</EnablePreviewFeatures>` and `<GenerateRequiresPreviewFeaturesAttribute>False</GenerateRequiresPreviewFeaturesAttribute>`. For consuming project I configured `<EnablePreviewFeatures>True</EnablePreviewFeatures>` and it didn't work.

This combination of flags is for super advanced scenarios where you want to ship a library that itself uses preview features but you don't want to force your consumers having to enable preview features. That's what we do internally in order to build, for example, `System.Runtime` itself. Application projects should never need that.

Let me take a look.

**Comment by buyaa-n** [comment]
> I followed two documents https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md and https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2252, but my understanding that putting `<EnablePreviewFeatures>true</EnablePreviewFeatures>` inside consuming .csproj doesn't work...

That should work, but it would not work if `<GenerateAssemblyInfo/>` property was set to false, because  MSBuild would add assembly-level `PreviewFeaturesAttribute` into the generated `AssemblyInfo.cs` file, but if the project is disabled generating this file the attribute could not be added. The workaround would be adding assembly-level `[assembly: System.Runtime.Versioning.RequiresPreviewFeaturesAttribute()]` attribute in your custom AssemblyInfo.cs file or anywhere in your project. Note that this will make your entire project/assembly `Preview assembly` and any usage from outside would cause the error 

> Last thing I tried was - in library .csproj file set `<EnablePreviewFeatures>True</EnablePreviewFeatures>` and `<GenerateRequiresPreviewFeaturesAttribute>False</GenerateRequiresPreviewFeaturesAttribute>`. For consuming project I configured `<EnablePreviewFeatures>True</EnablePreviewFeatures>` and it didn't work.


EDIT ~~Apparently, we added `<GenerateRequiresPreviewFeaturesAttribute>False</GenerateRequiresPreviewFeaturesAttribute>`  property to avoid adding the attribute in the assembly and also to avoid the analyzer warning within the project. But i don't see this is implemented in the analyzers, Filed [a bug](https://github.com/dotnet/roslyn-analyzers/issues/5722) in the analyzers repo~~ This is not a bug, it was by design
